### ISSUE TITLE
[BUG] DataFrame Boolean indexing does not properly handle the null mask but doesn't error

### ISSUE THREAD

**Issue by beckernick** [rapidsai/cudf#1475]
**Describe the bug**
Currently, Boolean indexing a dataframe containing null values at the dataframe level does not properly handle the null mask but does not cause the AssertionError that series level indexing with nulls causes. This can lead to users thinking they've correctly filtered their data, when they actually haven't.

To be specific, it appears that for every null value in a column, the post-filter column is cut short by one and filled with null values.

**Steps/Code to reproduce bug**
One null value (pandas first in all printed output):
```
import cudf
import pandas as pd

df = cudf.DataFrame({'a':[0,3,None,4,5]})
pdf = df.to_pandas()

print(pdf[pdf > 2])
print(df[df > 2])
     a
0  NaN
1  3.0
2  NaN
3  4.0
4  5.0
   a
0   
1  3
2  0
3  4
4   
```

Two null values:
```
import cudf
import pandas as pd

df = cudf.DataFrame({'a':[0, 3, None, 4, None, 5, 6, 7]})
pdf = df.to_pandas()

print(pdf[pdf > 2])
print(df[df > 2])
     a
0  NaN
1  3.0
2  NaN
3  4.0
4  NaN
5  5.0
6  6.0
7  7.0
   a
0   
1  3
2  0
3  4
4  0
5  5
6   
7   
```

Three null values:
```
import cudf
import pandas as pd

df = cudf.DataFrame({'a':[0, None, 3, None, 4, None, 5, 6, 7, 8]})
pdf = df.to_pandas()

print(pdf[pdf > 2])
print(df[df > 2])
     a
0  NaN
1  NaN
2  3.0
3  NaN
4  4.0
5  NaN
6  5.0
7  6.0
8  7.0
9  8.0
   a
0   
1  0
2  3
3  0
4  4
5  0
6  5
7   
8   
9   
```

**Expected behavior**
I expect dataframe level boolean indexing either to succeed correctly or raise an error.

**Environment details (please complete the following information):**
cudf 0.7 built from source. Commit SHA a87b8a8f2ea958b5315606d06412f7e3744a2fd2 (Apr 11).

**Additional context**
This issue may be resolved by the ongoing work for null support in 0.7.

**Comment by kkraus14** [comment]
Looks like this issue stems from there being nulls in the boolean mask column, where the following works:
```
import cudf
import pandas as pd

df = cudf.DataFrame({'a':[0, 3, None, 4, None, 5, 6, 7]})
pdf = df.to_pandas()

print(pdf[pdf > 2])
print(df[(df > 2).fillna(False)])
```

@harrism How would the libcudf `apply_boolean_mask` code react to the boolean mask having nulls in it? Ignore them and just read the data buffer?

**Comment by kkraus14** [comment]
Looks like this issue stems from there being nulls in the boolean masking dataframe. A `.fillna(False)` resolves the problem.

**Comment by cwharris** [comment]
numerical.py
```
...
    out = columnops.column_empty(row_count, dtype=out_dtype, masked=masked)
    # Call and fix null_count
    null_count = cpp_binops.apply_op(lhs, rhs, out, op)
...
````

I attempted to simply set `masked = False` when `op in ['gt', 'ge', 'lt', ...]`, but `cpp_binops` complained, expecting a mask. Instead, I added the following immediately after ward:
```
    if op in ['eq', 'ne', 'lt', 'le', 'gt', 'ge']: 
        out = out.replace(mask=None)
    else:
        out = out.replace(null_count=null_count)
```
This actually fixes the problem close to the source, but I'd still rather see CPP handle this, so I'm going to take a look at that.

When I ran tests for the fix above, certain binops comparison tests were failing. Specifically, `lt` comparisons involving nulls. These tests asserting that `None < x` where x is any number (maybe only positive? haven't explicitly checked yet). `None < x`, `x < None`, etc, should always evaluate to `False`, but we're asserting otherwise. I'm going to check this behavior against PANDAS.

`None < x` also mentioned in https://github.com/rapidsai/cudf/issues/1994#issuecomment-513379256

**Comment by cwharris** [comment]
This will be fixed with #2422. I'll look in to satisfying @beckernick's tests explicit.

**Comment by cwharris** [comment]
Fixed with #2422.
